### ISSUE TITLE
Cilium crashes when trying to `hubble observe`

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#14435]
Cilium development version during v1.10 dev cycle. 

```
# cilium version
Client: 1.9.90 9c803c18156c 2020-11-28T12:00:47-08:00 go version go1.15.6 linux/amd64
Daemon: 1.9.90 9c803c18156c 2020-11-28T12:00:47-08:00 go version go1.15.6 linux/amd64
# hubble version
hubble v0.7.1@HEAD-ba452ff compiled with go1.15.3 on linux/amd64
```

I was running several `hubble observe` commands in a row, for example:

```
$ hubble observe --namespace default -o json --last=3000
...
$ hubble observe --namespace default -o json --last=4000
rpc error: code = Unavailable desc = transport is closing
```

Then Cilium crashed:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x1e917a1]

goroutine 1044 [running]:
github.com/cilium/cilium/pkg/hubble/filters.sourcePod(0xc000c75c80, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cilium/cilium/pkg/hubble/filters/k8s.go:28 +0x41
github.com/cilium/cilium/pkg/hubble/filters.filterByNamespacedName.func1(0xc000c75c80, 0x200)
        /go/src/github.com/cilium/cilium/pkg/hubble/filters/k8s.go:59 +0x3e
github.com/cilium/cilium/pkg/hubble/filters.FilterFuncs.MatchAll(0xc0020806d0, 0x1, 0x1, 0xc000c75c80, 0x200)
        /go/src/github.com/cilium/cilium/pkg/hubble/filters/filters.go:40 +0x51
github.com/cilium/cilium/pkg/hubble/filters.BuildFilterList.func1(0xc000c75c80, 0x400)
        /go/src/github.com/cilium/cilium/pkg/hubble/filters/filters.go:127 +0x46
github.com/cilium/cilium/pkg/hubble/filters.FilterFuncs.MatchOne(0xc00129e660, 0x2, 0x2, 0xc000c75c80, 0x0)
        /go/src/github.com/cilium/cilium/pkg/hubble/filters/filters.go:55 +0x62
github.com/cilium/cilium/pkg/hubble/filters.Apply(0xc00129e660, 0x2, 0x2, 0x42ff0e0, 0x0, 0x0, 0xc000c75c80, 0x0)
        /go/src/github.com/cilium/cilium/pkg/hubble/filters/filters.go:34 +0x4d
github.com/cilium/cilium/pkg/hubble/observer.(*flowsReader).Next(0xc001139bc0, 0x2bd3980, 0xc001d2d8c0, 0x0, 0x0, 0x0)
        /go/src/github.com/cilium/cilium/pkg/hubble/observer/local_observer.go:400 +0x265
github.com/cilium/cilium/pkg/hubble/observer.(*LocalObserverServer).GetFlows(0xc0010dd9e0, 0xc000bb3f00, 0x2bec6e0, 0xc00129e640, 0x0, 0x0)
        /go/src/github.com/cilium/cilium/pkg/hubble/observer/local_observer.go:274 +0x753
github.com/cilium/cilium/api/v1/observer._Observer_GetFlows_Handler(0x25ce780, 0xc0010dd9e0, 0x2be69e0, 0xc000f26f00, 0x42ff0e0, 0xc001d3ce00)
        /go/src/github.com/cilium/cilium/api/v1/observer/observer_grpc.pb.go:121 +0x10b
google.golang.org/grpc.(*Server).processStreamingRPC(0xc001c56b60, 0x2bf8800, 0xc001e93500, 0xc001d3ce00, 0xc00182b4a0, 0x3d6d8c0, 0x0, 0x0, 0x0)
        /go/src/github.com/cilium/cilium/vendor/google.golang.org/grpc/server.go:1329 +0xcd8
google.golang.org/grpc.(*Server).handleStream(0xc001c56b60, 0x2bf8800, 0xc001e93500, 0xc001d3ce00, 0x0)
        /go/src/github.com/cilium/cilium/vendor/google.golang.org/grpc/server.go:1409 +0xc5c
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc001ea8b80, 0xc001c56b60, 0x2bf8800, 0xc001e93500, 0xc001d3ce00)
        /go/src/github.com/cilium/cilium/vendor/google.golang.org/grpc/server.go:746 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /go/src/github.com/cilium/cilium/vendor/google.golang.org/grpc/server.go:744 +0xa5
```

This is in a local microk8s cluster with the `examples/kubernetes/connectivity-check/connectivity-check-single-node.yaml` apps deployed in the default namespace.

**Comment by Rolinh** [comment]
My guess is that this issue is related to the recent addition of AgentEvent and the related filters support added in d7fa6e56f5c5618e61d08c2353f2392e9ce28f02. As now a `Flow` is not the only type on which filters apply, this `Event` method may return a non-typed nil, leading to the crash you observed:
```go
// GetFlow returns the decoded flow, or nil if there is no event
func (ev *Event) GetFlow() Flow {
        if ev == nil || ev.Event == nil {
                // returns typed nil so getter methods still work
                return (*pb.Flow)(nil)
        }
        if f, ok := ev.Event.(Flow); ok {
                return f
        }
        return nil
}
```

I believe that code that calls the getter for agent events might also be affected:
```go
// GetAgentEvent returns the decoded agent event, or nil if there is no event.
func (ev *Event) GetAgentEvent() AgentEvent {
        if ev == nil || ev.Event == nil {
                // returns typed nil so getter methods still work
                return (*pb.AgentEvent)(nil)
        }
        if f, ok := ev.Event.(AgentEvent); ok {
                return f
        }
        return nil
}
```

Callers of these 2 getter should now check for `nil` values as the event might not contain a flow or agent respectively.

TLDR: any code that does this (or a variant of it) may now crash:
```go
nodeName := ev.GetFlow().GetNodeName()
```

I don't think we want to return typed nil as there would be no way to tell whether the event is a flow or agent event. I think we need to check for `nil` at call site, Alternatively, we may change the methods signature to something like this:
```go
func (ev *Event) GetFlow() (f Flow, ok bool) {
        if ev == nil || ev.Event == nil {
                return nil, false
        }
        if f, ok := ev.Event.(Flow); ok {
                return f, true
        }
        return nil, false
}
```

**Comment by gandro** [comment]
@Rolinh Yup, I came to the same conclusion. My proposal to fix however would be to return the typed `return (*pb.Flow)(nil)` instead. It is nil-safe, so we do not have to adjust any call sites.

**Comment by Rolinh** [comment]
@gandro  The problem with this approach is that there would be no way to tell whether the event is a flow or agent event (except by doing a type check). Not sure if this is a problem in practice but if it is, I'd suggest adding `IsFlow() bool`/`IsAventEvent() bool` methods to `Event`.

**Comment by gandro** [comment]
Yes, I prefer the `IsFlow()`/`IsAgentEvent()` method over changing the `GetFlow()` signature. We have over two dozen call sites of `ev.GetFlow().GetFoo()` - it makes for very readable and nil-safe code that does the right thing even when the event is not a flow.

Having `GetFlow()` return two values will force all call sites to check if it really is a flow, when in fact most of them will work perfectly fine with the typed nil.

We should also re-evaluate as to why we are returning the `Flow` interface there. It's initial goal was to support extended flow implementations, which we never ended up doing. It might be easier to simply return a `*pb.Flow` instead.
